### ISSUE TITLE
AABBTree ray intersection clarification

### ISSUE THREAD

**Issue by HengguangZhou** [psarahdactyl/computer-graphics-bounding-volume-hierarchy#2]
I am confused by the Distance Query section, should we use bfs for AABBTree_ray_intersect.cpp? If so, what should we use as the distance for deciding the left or right child? I suppose the t value in ray_intersect_box should do the job, but the function is not returning it.

I implemented it following the pseudo code on the textbook, but I found the runtime not very satisfying, it's even slower than the brute force. 

Depth-first search result:
# Ray Triangle Mesh Intersection
  |V| 2050
  |F| 4096

  Firing 1000 rays...

  | Method      | Time in seconds |
  |:------------|----------------:|
  | brute force |   0.08437061310 |
  | build tree  |   0.58859753609 |
  | use tree    |   0.29415774345 |

**Comment by honglin-c** [comment]
Hi Hengguang, 
1. The Distance Query section refers to the program `./distances` for closest point queries where you'll need to implement BFS in `point_AABBTree_squared_distance.cpp`. (I admit the first sentence in Distance Query section is a bit misleading. :) )
2. The runtime for building the AABB tree seems to be much slower than what it should be. Maybe have a look at `AABBTree.cpp` about the construction process of the AABB tree?

**Comment by HengguangZhou** [comment]
Thank you for your reply.

I'm not sure why it is so slow, the code structure should be simple. Now I only find the longest axis by computing the difference between the largest X/Y/Z value on all those max_corner and the smallest X/Y/Z value on all the min_corner and then get the midpoint by adding the max and min X/Y/Z values and divide by two. Is this the right way to do this?

**Comment by HengguangZhou** [comment]
I got it working, there was a bug in the left-right split in my code. Much appreciated!

**Comment by honglin-c** [comment]
> Thank you for your reply.
> 
> I'm not sure why it is so slow, the code structure should be simple. Now I only find the longest axis by computing the difference between the largest X/Y/Z value on all those max_corner and the smallest X/Y/Z value on all the min_corner and then get the midpoint by adding the max and min X/Y/Z values and divide by two. Is this the right way to do this?

Actually the longest axis should be the one with the largest value among (max_corner(X)-min_corner(X)), (max_corner(Y)-min_corner(Y)) and (max_corner(Z)-min_corner(Z)).

**Comment by HengguangZhou** [comment]
Yes, my wording is kind of confusing, but I'm doing exactly the way you are saying here. Anyways, thank you very much!
